// @connect2ic/core/providers/dist/connect2ic-core-providers.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var providers_astrox_dist_connect2icCoreProvidersAstrox = require('../astrox/dist/connect2ic-core-providers-astrox.cjs.dev.js');
var providers_internetIdentity_dist_connect2icCoreProvidersInternetIdentity = require('../internet-identity/dist/connect2ic-core-providers-internet-identity.cjs.dev.js');
var providers_plugWallet_dist_connect2icCoreProvidersPlugWallet = require('../plug-wallet/dist/connect2ic-core-providers-plug-wallet.cjs.dev.js');
var providers_stoicWallet_dist_connect2icCoreProvidersStoicWallet = require('../stoic-wallet/dist/connect2ic-core-providers-stoic-wallet.cjs.dev.js');
var providers_infinityWallet_dist_connect2icCoreProvidersInfinityWallet = require('../infinity-wallet/dist/connect2ic-core-providers-infinity-wallet.cjs.dev.js');
var providers_nfid_dist_connect2icCoreProvidersNfid = require('../nfid/dist/connect2ic-core-providers-nfid.cjs.dev.js');
require('../../dist/classPrivateFieldGet-3c98eda8.cjs.dev.js');
require('@astrox/connection');
require('@astrox/connection/lib/esm/types');
require('@dfinity/auth-client');
require('@dfinity/agent');
require('ic-stoic-identity');

const defaultProviders = [providers_astrox_dist_connect2icCoreProvidersAstrox.AstroX, // EarthWallet,
providers_infinityWallet_dist_connect2icCoreProvidersInfinityWallet.InfinityWallet, providers_internetIdentity_dist_connect2icCoreProvidersInternetIdentity.InternetIdentity, providers_nfid_dist_connect2icCoreProvidersNfid.NFID, providers_plugWallet_dist_connect2icCoreProvidersPlugWallet.PlugWallet, providers_stoicWallet_dist_connect2icCoreProvidersStoicWallet.StoicWallet];
const walletProviders = [providers_astrox_dist_connect2icCoreProvidersAstrox.AstroX, providers_plugWallet_dist_connect2icCoreProvidersPlugWallet.PlugWallet];

exports.AstroX = providers_astrox_dist_connect2icCoreProvidersAstrox.AstroX;
exports.InternetIdentity = providers_internetIdentity_dist_connect2icCoreProvidersInternetIdentity.InternetIdentity;
exports.PlugWallet = providers_plugWallet_dist_connect2icCoreProvidersPlugWallet.PlugWallet;
exports.StoicWallet = providers_stoicWallet_dist_connect2icCoreProvidersStoicWallet.StoicWallet;
exports.InfinityWallet = providers_infinityWallet_dist_connect2icCoreProvidersInfinityWallet.InfinityWallet;
exports.NFID = providers_nfid_dist_connect2icCoreProvidersNfid.NFID;
exports.defaultProviders = defaultProviders;
exports.walletProviders = walletProviders;
